### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by iulianp26**
- new service which will process annotations received from Next Video Editor within the video content. See the Next Video Editor UPP support spec document: https://docs.google.com/a/ft.com/document/d/1ioJCmasEN2unBFnr1QcGj6Cf-QE0NKtgE2YVNYL3lVM/edit?usp=sharing
- tested in video cluster

### CODE REVIEW

File: annotations.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package main
+
+const relevanceURI = "http://api.ft.com/scoringsystem/FT-RELEVANCE-SYSTEM"
```
Comment (BerniVarga) [inline annotations.go]: You could put all the constants in one single const ( ) structure.

File: annotations.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	},
+}
+
+type annHandler struct {
+	videoUUID     string
+	transactionID string
+}
+
+func (h annHandler) createAnnotations(nextAnns []annotation) ConceptSuggestion {
+	var suggestions = make([]suggestion, 0)
```
Comment (BerniVarga) [inline annotations.go]: I don't think the initialisation at this point has any advantages, you could use a simple zero value var declaration

File: annotations_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	return annotation{
+		thingID:     thingID,
+		predicate: predicate,
+	}
+}
+
+func newSuggestion(thingID string, predicate string) suggestion {
+	thing := thing{
+		ID:        thingID,
+		PrefLabel: "",
```
Comment (BerniVarga) [inline annotations_test.go]: Zero value of a string type is "", so the prefLabel initialisation can be omitted here. Might be similar for types - depending on how you use it through your code..

File: annotations.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	if len(suggestions) == 0 {
+		logger.videoEvent(h.transactionID, h.videoUUID, "No annotation could be mapped for the video")
+	}
+
+	return ConceptSuggestion{UUID: h.videoUUID, Suggestions: suggestions}
+}
+
+func (h annHandler) createAnnotation(nextAnn annotation) suggestion {
```
Comment (BerniVarga) [inline annotations.go]: I would use new instead of create - more like a golang standard

File: annotations.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		},
+	},
+}
+
+type annHandler struct {
+	videoUUID     string
+	transactionID string
+}
+
+func (h annHandler) createAnnotations(nextAnns []annotation) ConceptSuggestion {
```
Comment (BerniVarga) [inline annotations.go]: Does it make sense to create these methods over the handler? I think they could be simple functions with parameters (easier to read). Eventually, a context parameter could be used instead - for logging.

File: app.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+	logger.serviceStartedEvent(sc.asMap())
+
+	err := http.ListenAndServe(":"+sc.appPort, r)
+	if err != nil {
+		logrus.Fatalf("Unable to start server: %v", err)
+	}
+}
+
+func consumeUntilSigterm(messageConsumer *consumer.MessageConsumer, config consumer.QueueConfig) {
```
Comment (BerniVarga) [inline app.go]: MessageConsumer is an interface, it is fine to define it without the pointer ref

File: applogger.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
+package main
+
+import (
+	"github.com/Sirupsen/logrus"
+)
```
Comment (BerniVarga) [inline applogger.go]: If it's a single import line, no need for the brackets

File: circle.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+dependencies:
+  pre:
+    - go get github.com/axw/gocov/gocov; go get github.com/matm/gocov-html; go get -u github.com/jstemmer/go-junit-report
```
Comment (BerniVarga) [inline circle.yml]: We don't use gocov recently, you can remove that dependency

File: healthchecks.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		Name:             "MessageQueueProxyReachable",
+		PanicGuide:       h.panicGuide,
+		Severity:         1,
+		TechnicalSummary: "Message queue proxy is not reachable/healthy",
+		Checker:          h.checkAggregateMessageQueueProxiesReachable,
+	}
+}
+
+func (h *healthCheck) checkAggregateMessageQueueProxiesReachable() (string, error) {
+	errMsg := ""
```
Comment (BerniVarga) [inline healthchecks.go]: `var errMsg string`
It's the same as yours, but i think it looks prettier. Sure, it's also a question of preference. :)

File: mapper.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+const videoUUIDField = "id"
```
Comment (BerniVarga) [inline mapper.go]: Please group consst in a single bracket

File: queuehandler.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+const nextVideoOrigin = "http://cmdb.ft.com/systems/next-video-editor"
+const dateFormat = "2006-01-02T03:04:05.000Z0700"
+const generatedMsgType = "concept-suggestions"
+
+type queueHandler struct {
+	sc              serviceConfig
+	consumerConfig  consumer.QueueConfig
+	producerConfig  producer.MessageProducerConfig
+	messageConsumer *consumer.MessageConsumer
```
Comment (BerniVarga) [inline queuehandler.go]: You could remove the pointer reference, since it is an interface (same for producer)

File: queuehandler.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	messageProducer *producer.MessageProducer
+}
+
+func (h *queueHandler) init() {
+	messageProducer := producer.NewMessageProducer(h.producerConfig)
+	h.messageProducer = &messageProducer
+
+	httpCl := http.Client{}
```
Comment (BerniVarga) [inline queuehandler.go]: Would it make sense to use some config for the client object injected from outside? (for timeout, keep alive, or max connection values?) You could create the client config in your main, and pass that client through all of your code

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  && REPOSITORY="repository=$(git config --get remote.origin.url)" \
+  && REVISION="revision=$(git rev-parse HEAD)" \
+  && BUILDER="builder=$(go version)" \
+  && LDFLAGS="-X '"${BUILDINFO_PACKAGE}$VERSION"' -X '"${BUILDINFO_PACKAGE}$DATETIME"' -X '"${BUILDINFO_PACKAGE}$REPOSITORY"' -X '"${BUILDINFO_PACKAGE}$REVISION"' -X '"${BUILDINFO_PACKAGE}$BUILDER"'" \
+  && echo $LDFLAGS \
+  && mkdir -p $GOPATH/src/${REPO_PATH} \
+  && mv * $GOPATH/src/${REPO_PATH} \
+  && cd $GOPATH/src/${REPO_PATH} \
+  && go get -t ./... \
+  && go test ./... \
```
Comment (tosan88) [inline Dockerfile]: Could you please include the `-race` flag?

File: annotations.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+	if len(suggestions) == 0 {
+		logger.videoEvent(h.transactionID, h.videoUUID, "No annotation could be mapped for the video")
+	}
+
+	return ConceptSuggestion{UUID: h.videoUUID, Suggestions: suggestions}
+}
+
+func (h annHandler) createAnnotation(nextAnn annotation) suggestion {
+	thing := thing{
```
Comment (tosan88) [inline annotations.go]: It's not a problem here, but could be in other cases, so I think it's a useful FYI: `declaration of thing shadows declaration at annotations.go:20 (vetshadow)` :)

File: mapper.go [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+		obj, ok = objI.(map[string]interface{})
+		if !ok {
+			return nil, wrongFieldTypeError("object array", key, objArrayI)
+		}
+		result = append(result, obj)
+	}
+	return result, nil
+}
+
+func getBoolField(key string, obj map[string]interface{}) (bool, error) {
```
Comment (tosan88) [inline mapper.go]: I think this function was for the primary flag, which we don't care anymore. So it could be removed

File: mapper.go [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (vm *videoMapper) mapNextVideoAnnotations() ([]byte, string, error) {
+	videoUUID, err := getRequiredStringField(videoUUIDField, vm.unmarshalled)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if vm.isDeleteEvent() {
+		logger.videoMapEvent(vm.tid, videoUUID, fmt.Sprint("Ignoring delete Next video message."))
```
Comment (tosan88) [inline mapper.go]: :+1:

File: .gitignore [hunk 17]
```diff
@@ -1,24 +1,7 @@
...
-*.exe
-*.test
-*.prof
+upp-next-video-annotations-mapper.exe
+*.log
+*.html
+*coverage.*
+*.iml
+.idea/
+upp-next-video-annotations-mapper
```
Comment (tosan88) [inline .gitignore]: Maybe worth removing the `upp-next-video-annotations-mapper.exe` line and adding a `upp-next-video-annotations-mapper*` regex match

File: healthchecks_test.go [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/gorilla/handlers"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+const (
+	_            = iota
```
Comment (tosan88) [inline healthchecks_test.go]: Why is this needed?

File: annotations.go (hunk 1 above)
Comment (iulianp26) [inline annotations.go]: Done.

File: annotations.go (hunk 2 above)
Comment (iulianp26) [inline annotations.go]: Done.

File: annotations_test.go (hunk 3 above)
Comment (iulianp26) [inline annotations_test.go]: Done.

File: annotations.go (hunk 4 above)
Comment (iulianp26) [inline annotations.go]: Done.

File: annotations.go (hunk 5 above)
Comment (iulianp26) [inline annotations.go]: Done.

File: app.go (hunk 6 above)
Comment (iulianp26) [inline app.go]: Done.

File: applogger.go (hunk 7 above)
Comment (iulianp26) [inline applogger.go]: Done.

File: circle.yml (hunk 8 above)
Comment (iulianp26) [inline circle.yml]: Done.

File: healthchecks.go (hunk 9 above)
Comment (iulianp26) [inline healthchecks.go]: Done.

File: mapper.go (hunk 10 above)
Comment (iulianp26) [inline mapper.go]: Done.

File: queuehandler.go (hunk 11 above)
Comment (iulianp26) [inline queuehandler.go]: Done.

File: queuehandler.go (hunk 12 above)
Comment (iulianp26) [inline queuehandler.go]: Done.

File: Dockerfile (hunk 13 above)
Comment (iulianp26) [inline Dockerfile]: Done.

File: annotations.go (hunk 14 above)
Comment (iulianp26) [inline annotations.go]: Done.

File: mapper.go (hunk 15 above)
Comment (iulianp26) [inline mapper.go]: Done.

File: healthchecks_test.go (hunk 18 above)
Comment (iulianp26) [inline healthchecks_test.go]: Correct, I've removed it.

File: healthchecks.go [hunk 19]
```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	fthealth "github.com/Financial-Times/go-fthealth/v1a"
```
Comment (radutudor) [inline healthchecks.go]: latest healthcheck version is `v1_1`

File: vendor/vendor.json [hunk 20]
```diff
@@ -0,0 +1,99 @@
+{
+	"comment": "",
+	"ignore": "test",
+	"package": [
+		{
+			"checksumSHA1": "4gy/7n/lgtdi5efuHXwFcIe1ZI8=",
+			"path": "github.com/Financial-Times/go-fthealth/v1a",
+			"revision": "f9074331b3ab48bca0e6d0afc211a673d88ac925",
```
Comment (bogdanguranda) [inline vendor/vendor.json]: I think this could be an improvement task also: 
We encourage use of specific version tags for vendoring instead of only a revision number (as it is more human readable and useful), at least for FT repos, so for this case you could run this to update with the latest release tag of this library: govendor fetch github.com/Financial-Times/go-fthealth/v1a@0.2.0

File: app.go [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+	serviceName string
+	appPort     string
+}
+
+func main() {
+	app := cli.App("upp-next-video-annotations-mapper", serviceDescription)
+	serviceName := app.StringOpt("app-name", "next-video-annotations-mapper", "The name of this service")
+	appPort := app.String(cli.StringOpt{
+		Name:   "app-port",
+		Value:  "8084",
```
Comment (iulianp26) [inline app.go]: It remains like that from when doing local tests on the map endpoint with tunneling enabled and default 8080 couldn't be used as request on localhost:8080 get forwarded to the cluster. I will change it to 8080.
